### PR TITLE
Fix loading dynamic functions/classes when using code_search_path

### LINKED ISSUE #12834: Remove load_code_from_local
### Describe your feature request
Currently, the python code can be loaded from gcs or local, but can't can loaded from both side. It would be great if we can try load code from local first, and load from gcs if we can't load cpode from local.

### PR DESCRIPTION
**Pull request by jypeng28**
## Why are these changes needed?
When `code_search_path` is set,  we will always load functions and actor classes from local code. Thus dynamically defined functions and classes cannot be loaded. 

This PR added a mixed data-load mechanism. When loading a function/class,  the worker tries to load it from local first. If it fails, then load it from GCS. Also, only dynamic functions/classes will be exported to GCS, in order to save GCS space.

## Related issue number

closes #12834 

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/_private/function_manager.py [hunk 1]
```diff
@@ -124,15 +118,37 @@ def compute_collision_identifier(self, function_or_class):
         # Return a hash of the identifier in case it is too large.
         return hashlib.sha1(collision_identifier.encode("utf-8")).digest()
 
+    def check_local(self,module_name,function_class_name):
```
Comment (raulchen) [inline python/ray/_private/function_manager.py:121]: better name this `load_function_or_class_from_local`, as it returns the loaded object. 
`check_sth` usually returns a boolean
```
    def load_function_or_class_from_local(self, module_name, function_or_class_name):
```

File: python/ray/_private/function_manager.py [hunk 2]
```diff
@@ -124,15 +118,37 @@ def compute_collision_identifier(self, function_or_class):
...
+        if function_descriptor.is_actor_method():
+            return
+        function_id = function_descriptor.function_id
+        if (function_id in self._function_execution_info):
+            return
+        module_name, function_name = (
+            function_descriptor.module_name,
+            function_descriptor.function_name,
+        )
+        if self.check_local(module_name,function_name):
```
Comment (raulchen) [inline python/ray/_private/function_manager.py:150]: Please add some comments in each if statement. E.g., "If the function is an actor method, no need to export it to GCS".

File: python/ray/_private/function_manager.py [hunk 3]
```diff
@@ -227,61 +243,55 @@ def get_execution_info(self, job_id, function_descriptor):
...
+            info = self._function_execution_info[function_id]
         except KeyError as e:
             message = ("Error occurs in get_execution_info: "
                        "job_id: %s, function_descriptor: %s. Message: %s" %
                        (job_id, function_descriptor, e))
             raise KeyError(message)
         return info
 
-    def _load_function_from_local(self, job_id, function_descriptor):
+    def _load_function_from_local(self, function_descriptor):
```
Comment (raulchen) [inline python/ray/_private/function_manager.py:271]: This method should better just return a boolean. The exception is no longer useful.

File: python/ray/_private/function_manager.py [hunk 4]
```diff
@@ -227,61 +243,55 @@ def get_execution_info(self, job_id, function_descriptor):
...
-            job_id = ray.JobID.nil()
+
+        try:
+            # Try to load function from local code first.
             if not function_descriptor.is_actor_method():
-                self._load_function_from_local(job_id, function_descriptor)
-        else:
+                self._load_function_from_local(function_descriptor)
+        except:
             # Load function from GCS.
```
Comment (raulchen) [inline python/ray/_private/function_manager.py:252]: Add comment: "If the function doesn't exist in local code, try loading it from GCS."

File: python/ray/_private/function_manager.py [hunk 5]
```diff
@@ -227,61 +243,55 @@ def get_execution_info(self, job_id, function_descriptor):
...
             raise KeyError(message)
         return info
 
-    def _load_function_from_local(self, job_id, function_descriptor):
+    def _load_function_from_local(self, function_descriptor):
         assert not function_descriptor.is_actor_method()
         function_id = function_descriptor.function_id
-        if (job_id in self._function_execution_info
-                and function_id in self._function_execution_info[job_id]):
+        if (function_id in self._function_execution_info):
```
Comment (raulchen) [inline python/ray/_private/function_manager.py:274]: nit: remove the brackets.

File: python/ray/tests/test_basic_2.py [hunk 6]
```diff
@@ -720,6 +720,26 @@ def get(self):
         assert b"OK" in output
 
 
+def test_use_dynamic_function_and_class(ray_start_regular_shared):
```
Comment (raulchen) [inline python/ray/tests/test_basic_2.py:723]:
```
def test_use_dynamic_function_and_class(ray_start_regular_shared):
    """Test use dynamically defined functions and classes for remote tasks and actors.
         See https://github.com/ray-project/ray/issues/12834.
    """
```

**Comment by raulchen** [comment]
@rkooo567 @iycheng @architkulkarni  Please let us know if this new mechanism is not compatible with your usage.

File: python/ray/_private/function_manager.py (hunk 1 above)
Comment (kfstorm) [inline python/ray/_private/function_manager.py:121]: I'm a bit worried about the performance of this function. Can we add a local cache here?

File: python/ray/_private/function_manager.py [hunk 7]
```diff
@@ -124,15 +118,37 @@ def compute_collision_identifier(self, function_or_class):
...
             remote_function: the RemoteFunction object.
         """
-        if self._worker.load_code_from_local:
-            return
 
+        function_descriptor = remote_function._function_descriptor
+        if function_descriptor.is_actor_method():
+            return
+        function_id = function_descriptor.function_id
+        if (function_id in self._function_execution_info):
```
Comment (raulchen) [inline python/ray/_private/function_manager.py:144]: remove this

File: python/ray/_private/function_manager.py (hunk 5 above)
Comment (raulchen) [reply python/ray/_private/function_manager.py:274]: move this check to get_execution_info

File: python/ray/_private/function_manager.py (hunk 1 above)
Comment (raulchen) [reply python/ray/_private/function_manager.py:121]: Just checked, this should be fine. Loading already have a cache. Exporting will be only done once.

File: python/ray/_private/function_manager.py [hunk 8]
```diff
@@ -124,18 +119,38 @@ def compute_collision_identifier(self, function_or_class):
...
         """Pickle a remote function and export it to redis.
 
         Args:
             remote_function: the RemoteFunction object.
         """
-        if self._worker.load_code_from_local:
-            return
 
+        function_descriptor = remote_function._function_descriptor
+        #If the function is an actor method, no need to export it to GCS#
```
Comment (raulchen) [inline python/ray/_private/function_manager.py:142]: remove the trailing #

File: python/ray/_private/function_manager.py [hunk 9]
```diff
@@ -124,18 +119,38 @@ def compute_collision_identifier(self, function_or_class):
...
+        if function_descriptor.is_actor_method():
+            return
+        module_name, function_name = (
+            function_descriptor.module_name,
+            function_descriptor.function_name,
+        )
+        #If the function is local, no need to export it to GCS#
+        if self.load_function_or_class_from_local(module_name,function_name):
```
Comment (raulchen) [inline python/ray/_private/function_manager.py:150]: Add ` != None`

File: python/ray/_private/function_manager.py [hunk 10]
```diff
@@ -124,18 +119,38 @@ def compute_collision_identifier(self, function_or_class):
...
+        """Try to load a function or class in the module from local."""
+        module = importlib.import_module(module_name)
+        parts = [part for part in function_or_class_name.split(".") if part]
+        object = module
+        try:
+            for part in parts:
+                object = getattr(object, part)
+            return object
+        except:
+            return
```
Comment (raulchen) [inline python/ray/_private/function_manager.py:132]: return None

File: python/ray/_private/function_manager.py [hunk 11]
```diff
@@ -392,18 +402,15 @@ def load_actor_class(self, job_id, actor_creation_function_descriptor):
...
-                job_id = ray.JobID.nil()
-                # Load actor class from local code.
-                actor_class = self._load_actor_class_from_local(
+            # Try to load actor class from local code first.
+            actor_class = self._load_actor_class_from_local(
                     job_id, actor_creation_function_descriptor)
-            else:
-                # Load actor class from GCS.
+            if actor_class is None:
+                # Load actor class from GCS if it fails.
```
Comment (raulchen) [inline python/ray/_private/function_manager.py:409]: if loading local fails ("it" is a bit ambiguous)

**Comment by fyrestone** [comment]
I noticed that there are lots of code format changes, can you check your format tool versions? It's better to keep the tool version the same as ray.

File: python/ray/_private/function_manager.py [hunk 12]
```diff
@@ -118,29 +111,52 @@ def compute_collision_identifier(self, function_or_class):
...
+        collision_identifier = (function_or_class.__name__ + ":" +
+                                string_file.getvalue())
 
         # Return a hash of the identifier in case it is too large.
         return hashlib.sha1(collision_identifier.encode("utf-8")).digest()
 
+    def load_function_or_class_from_local(self, module_name,
+                                          function_or_class_name):
+        """Try to load a function or class in the module from local."""
+        module = importlib.import_module(module_name)
```
Comment (fyrestone) [inline python/ray/_private/function_manager.py:124]: This line may raises exceptions.

If the target object is defined in `__main__`, then we get the module name from it's code object, please refer to https://github.com/ray-project/ray/blob/master/python/ray/includes/function_descriptor.pxi#L288. But, the `opentelemetry` will patch a code object from `tracing_helper`, then you will get a incorrect module name.

```python
import ray
ray.init()
@ray.remote
def foo(): return 'foo'

foo.remote()  # Generate the function descriptor
print(foo._function_descriptor)  # Get an incorrect function descriptor
# output: {type=PythonFunctionDescriptor, module_name=tracing_helper, class_name=, function_name=foo, function_hash=47d95eadcfdd01484b125defd1b5b1403994dcb69422e90bfac573df}
```

Comment (fyrestone) [reply python/ray/_private/function_manager.py:124]: I think we need a better way to extract the module name from target, then add a try to handle the import error.

File: python/ray/tests/test_basic_2.py [hunk 13]
```diff
@@ -720,6 +720,29 @@ def get(self):
...
+        return "OK"
+
+    @ray.remote
+    class Foo:
+        def foo(self):
+            return "OK"
+
+    assert ray.get(foo.remote()) == "OK"
+    foo_actor = Foo.remote()
+    assert ray.get(foo_actor.foo.remote()) == "OK"
```
Comment (rkooo567) [inline python/ray/tests/test_basic_2.py:743]: This doesn't test if dynamic functions are actually stored in GCS. Can you actually use the Redis client and parse the function table to verify that?

File: python/ray/_private/function_manager.py [hunk 14]
```diff
@@ -124,15 +124,38 @@ def compute_collision_identifier(self, function_or_class):
...
+            if function_descriptor.is_actor_method():
+                return
+            module_name, function_name = (
+                function_descriptor.module_name,
+                function_descriptor.function_name,
+            )
+            # If the function is dynamic, we still export it to GCS
+            # even if load_code_from_local is set True.
+            if self.load_function_or_class_from_local(module_name,
+                                                      function_name):
```
Comment (raulchen) [inline python/ray/_private/function_manager.py:157]: add `!= None`

File: python/ray/_private/function_manager.py [hunk 15]
```diff
@@ -231,9 +254,17 @@ def get_execution_info(self, job_id, function_descriptor):
             # Load function from local code.
             # Currently, we don't support isolating code by jobs,
             # thus always set job ID to NIL here.
-            job_id = ray.JobID.nil()
+            job_id_local = ray.JobID.nil()
             if not function_descriptor.is_actor_method():
-                self._load_function_from_local(job_id, function_descriptor)
+                try:
+                    self._load_function_from_local(job_id_local,
```
Comment (raulchen) [inline python/ray/_private/function_manager.py:260]: we should still let `_load_function_from_local` use `load_function_or_class_from_local`

File: python/ray/_private/function_manager.py [hunk 16]
```diff
@@ -344,7 +375,14 @@ def _publish_actor_class_to_key(self, key, actor_class_info):
     def export_actor_class(self, Class, actor_creation_function_descriptor,
                            actor_method_names):
         if self._worker.load_code_from_local:
-            return
+            module_name, class_name = (
+                actor_creation_function_descriptor.module_name,
+                actor_creation_function_descriptor.class_name)
+            # If the class is dynamic, we still export it to GCS
+            # even if load_code_from_local is set True.
+            if self.load_function_or_class_from_local(module_name, class_name):
```
Comment (raulchen) [inline python/ray/_private/function_manager.py:383]: same here `!= None`

File: python/ray/worker.py [hunk 17]
```diff
@@ -494,12 +494,22 @@ def get_gpu_ids():
     worker = global_worker
     worker.check_connected()
 
+    if worker.mode != WORKER_MODE:
```
Comment (raulchen) [inline python/ray/worker.py:497]: this change is unrelated?

File: python/ray/tests/test_basic_2.py [hunk 18]
```diff
@@ -720,6 +720,46 @@ def get(self):
...
+        b"RemoteFunction:" + ray.worker.global_worker.current_job_id.binary() +
+        b":" + f._function_descriptor.function_id.binary())
+    (job_id_str, function_id_str, function_name, serialized_function, module,
+     max_calls) = ray.worker.global_worker.redis_client.hmget(
+         key_, [
+             "job_id", "function_id", "function_name", "function", "module",
+             "max_calls"
+         ])
+    assert f._function_descriptor.function_id == ray.FunctionID(
+        function_id_str)
```
Comment (architkulkarni) [inline python/ray/tests/test_basic_2.py:758]: Could you please add a comment explaining what this is testing?  

It looks like we're only using `function_id_str` here, is there a reason we're requesting all those other fields?  Is it to make sure they were set correctly?

File: python/ray/_private/function_manager.py [hunk 19]
```diff
@@ -124,15 +114,37 @@ def compute_collision_identifier(self, function_or_class):
...
-
         Args:
             remote_function: the RemoteFunction object.
         """
         if self._worker.load_code_from_local:
-            return
-
+            function_descriptor = remote_function._function_descriptor
+            if function_descriptor.is_actor_method():
+                return
```
Comment (architkulkarni) [inline python/ray/_private/function_manager.py:138]: Could you please add a comment explaining why we return in this case?

File: python/ray/tests/test_basic_2.py [hunk 20]
```diff
@@ -720,6 +720,46 @@ def get(self):
...
+    def foo1():
+        @ray.remote
+        def foo2():
+            return "OK"
+
+        return foo2
+
+    @ray.remote
+    class Foo:
+        @ray.method(num_returns=1)
```
Comment (architkulkarni) [inline python/ray/tests/test_basic_2.py:742]: Out of curiosity, what is this line for?  Does it work without this line?

File: python/ray/_private/function_manager.py (hunk 12 above)
Comment (raulchen) [reply python/ray/_private/function_manager.py:120]: talked offline, this is an existing issue. we can handle it in the next PR.

File: python/ray/_private/function_manager.py [hunk 21]
```diff
@@ -219,21 +231,30 @@ def f(*args, **kwargs):
...
             A FunctionExecutionInfo object.
         """
         if self._worker.load_code_from_local:
             # Load function from local code.
             # Currently, we don't support isolating code by jobs,
             # thus always set job ID to NIL here.
-            job_id = ray.JobID.nil()
-            if not function_descriptor.is_actor_method():
-                self._load_function_from_local(job_id, function_descriptor)
+            function_id = function_descriptor.function_id
```
Comment (raulchen) [inline python/ray/_private/function_manager.py:244]: This would be clearer
```
if already in _function_execution_info:
   return info
if load_code_from_local:
   if not actor method:
       info = self._load_function_from_local()
       if info is not None:
           return info
# Load from GCS
...
   
```

File: python/ray/tests/test_basic_2.py [hunk 22]
```diff
@@ -720,6 +720,46 @@ def get(self):
...
+    key_ = (
+        b"RemoteFunction:" + ray.worker.global_worker.current_job_id.binary() +
+        b":" + f._function_descriptor.function_id.binary())
+    (job_id_str, function_id_str, function_name, serialized_function, module,
+     max_calls) = ray.worker.global_worker.redis_client.hmget(
+         key_, [
+             "job_id", "function_id", "function_name", "function", "module",
+             "max_calls"
+         ])
+    assert f._function_descriptor.function_id == ray.FunctionID(
```
Comment (raulchen) [inline python/ray/tests/test_basic_2.py:757]: 1) Just checking that the key exists should be enough.
2) Also check the actor class.
3) Also please add a comment that the key format should be the same as "FunctionManager.export"

File: python/ray/tests/test_basic_2.py (hunk 20 above)
Comment (Jingyu-Peng) [reply python/ray/tests/test_basic_2.py:742]: Yes, it's ok to drop @ray.method(num_returns=1).

File: python/ray/tests/test_basic_2.py [hunk 23]
```diff
@@ -720,6 +720,49 @@ def get(self):
         assert b"OK" in output
 
 
+@pytest.mark.skipif(client_test_enabled(), reason="init issue")
```
Comment (raulchen) [inline python/ray/tests/test_basic_2.py:723]: `reason="JobConfig doesn't work in client mode"

File: python/ray/tests/test_basic_2.py [hunk 24]
```diff
@@ -720,6 +720,49 @@ def get(self):
...
+        return foo2
+
+    @ray.remote
+    class Foo:
+        @ray.method(num_returns=1)
+        def foo(self):
+            return "OK"
+
+    # To check if the dynamic function is stored in GCS
+    # Keep the key of function the same as export.
```
Comment (raulchen) [inline python/ray/tests/test_basic_2.py:747]: Move this comment after line 749.
```
    # Check that the dynamic function is exported to GCS.
    # Note, the key format should be kept in sync with the one in `FunctionManager.export`.
```

**Comment by architkulkarni** [comment]
> @rkooo567 @iycheng @architkulkarni Please let us know if this new mechanism is not compatible with your usage.

I think it should be fine.  `runtime_env["working_dir"]` is just a way to push code to all nodes in the cluster and have the workers start in these directories, and my understanding is that `load_code_from_local` is one way of starting tasks and actors defined in these files, so these should be compatible.

File: python/ray/_private/function_manager.py [hunk 25]
```diff
@@ -219,80 +229,74 @@ def f(*args, **kwargs):
...
         """
         if self._worker.load_code_from_local:
             # Load function from local code.
-            # Currently, we don't support isolating code by jobs,
-            # thus always set job ID to NIL here.
-            job_id = ray.JobID.nil()
+            function_id = function_descriptor.function_id
+            # If the function has already been loaded,
+            # There's no need to load again
+            if function_id in self._function_execution_info:
```
Comment (raulchen) [inline python/ray/_private/function_manager.py:243]: this can be moved before `if self._worker.load_code_from_local:`

File: python/ray/tests/test_basic_2.py [hunk 26]
```diff
@@ -720,6 +720,50 @@ def get(self):
...
+    @ray.remote
+    class Foo:
+        @ray.method(num_returns=1)
+        def foo(self):
+            return "OK"
+
+    f = foo1()
+    assert ray.get(f.remote()) == "OK"
+    # To check if the dynamic function is stored in GCS
+    # Keep the key of function the same as export.
```
Comment (raulchen) [inline python/ray/tests/test_basic_2.py:750]:
```
Check whether the dynamic function is exported to GCS.
Note, the key format should be kept the same as in `FunctionManager.export`.
```

**Review by raulchen — APPROVED** [review]

**Review by fyrestone — APPROVED** [review]
LGTM

**Comment by fyrestone** [comment]
Please fix the lint.
